src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -248,52 +292,174 @@ public static OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int b
                 destIndex += 2;
             }
             else
             {
                 if (i0 < 0)
                     goto InvalidExit;
-                Unsafe.Add(ref bufferBytes, destIndex) = (byte)(i0 >> 16);
+
+                Unsafe.Add(ref bufferBytes, (IntPtr)destIndex) = (byte)(i0 >> 16);
                 destIndex += 1;
             }
 
         DoneExit:
-            bytesWritten = destIndex;
+            bytesWritten = (int)destIndex;
             return OperationStatus.Done;
 
         InvalidExit:
-            bytesWritten = destIndex;
+            bytesWritten = (int)destIndex;
             return OperationStatus.InvalidData;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Avx2Decode(ref byte src, ref byte destBytes, int sourceLength, int destLength, ref uint sourceIndex, ref uint destIndex)
+        {
+            ref byte srcStart = ref src;
+            ref byte destStart = ref destBytes;
+            ref byte simdSrcEnd = ref Unsafe.Add(ref src, (IntPtr)((uint)sourceLength - 45 + 1));
+
+            // The JIT won't hoist these "constants", so help him
+            Vector256<sbyte> lutHi = s_avxDecodeLutHi;
+            Vector256<sbyte> lutLo = s_avxDecodeLutLo;
+            Vector256<sbyte> lutShift = s_avxDecodeLutShift;
+            Vector256<sbyte> mask2F = s_avxDecodeMask2F;
+            Vector256<sbyte> shuffleConstant0 = Vector256.Create(0x01400140).AsSByte();
+            Vector256<short> shuffleConstant1 = Vector256.Create(0x00011000).AsInt16();
+            Vector256<sbyte> shuffleVec = s_avxDecodeShuffleVec;
+            Vector256<int> permuteVec = s_avxDecodePermuteVec;
+
+            //while (remaining >= 45)
+            do
+            {
+                AssertRead<Vector256<sbyte>>(ref src, ref srcStart, sourceLength);
+                Vector256<sbyte> str = Unsafe.As<byte, Vector256<sbyte>>(ref src);
+
+                Vector256<sbyte> hiNibbles = Avx2.And(Avx2.ShiftRightLogical(str.AsInt32(), 4).AsSByte(), mask2F);
+                Vector256<sbyte> loNibbles = Avx2.And(str, mask2F);
+                Vector256<sbyte> hi = Avx2.Shuffle(lutHi, hiNibbles);
+                Vector256<sbyte> lo = Avx2.Shuffle(lutLo, loNibbles);
+                Vector256<sbyte> zero = Vector256<sbyte>.Zero;
+
+                // https://github.com/dotnet/coreclr/issues/21247
+                if (Avx2.MoveMask(Avx2.CompareGreaterThan(Avx2.And(lo, hi), zero)) != 0)
+                    break;
+
+                Vector256<sbyte> eq2F = Avx2.CompareEqual(str, mask2F);
+                Vector256<sbyte> shift = Avx2.Shuffle(lutShift, Avx2.Add(eq2F, hiNibbles));
+                str = Avx2.Add(str, shift);
+
+                Vector256<short> merge_ab_and_bc = Avx2.MultiplyAddAdjacent(str.AsByte(), shuffleConstant0);
+                Vector256<int> @out = Avx2.MultiplyAddAdjacent(merge_ab_and_bc, shuffleConstant1);
```

Comment:
nit: would be nice to name this something other than `@out` (maybe `output`?)

Reply:
Done -- will be updated with the next push.